Add an unhandledError callback

dnfield opened the pull request:
A published version with this would help  flutter/flutter#28531

We ran into this scenario when a cast (that I believe worked in an earlier Dart version) no longer was acceptable.

/cc @kevmoo @liyuqian

dnfield commented:
Alternatively, we might be able to just print the exception to stderr...

kevmoo commented:
Thoughts @nex3 ?

dnfield commented:
Actually the more I think about this he more I'm convinced rethrow is right.  It may disrupt clients who are missing stuff today, but that means they may be broken today and not even realize it.

liyuqian commented:
I agree that `rethrow` would be better for our overall quality. But be aware that Flutter is a client that's missing stuff today. I'm pretty sure that the flutter_gallery__transition_perf will crash if we `rethrow`. If infinite resources are given, I'd choose `rethrow` and fix all our breaking tests.

dnfield commented:
If we're really concerned about it, we could wrap the calls in our own exception swallower until we fix it right?

nex3 commented:
> Thoughts @nex3 ?

I neither maintain nor use this package anymore :woman_shrugging:. I will say that throwing an error due to an invalid message sent spontaneously by an unknown and untrusted third party *seems* like it could be a vector for DOS attacks against open JSON-RPC 2.0 servers. Perhaps it would be more robust to have the constructor take an optional `void onHandlerError(error, StackTrace)` callback?

dnfield commented:
That would work for me.  My main concern is that right now, this is swallowing exceptions that hide bugs or cause flakes in our tests.  I can refactor this PR to use the onHandleError approach.

dnfield commented:
Updated to use the callback mechanism and added a test.  I also updated the CHANGELOG and bumped the version number.  It seems like the CHANGELOG is missing an entry for 2.0.9 though, so I'm not sure what to do about that.

dnfield commented:
And to clarify further, this circumstance is arising when a callback provided to the server throws an exception.  The server ends up swallowing that exception instead of allowing it to be thrown.  The downside of this is that we likely have errors in our code using this that we've never caught until now.

liyuqian commented on test/peer_test.dart:
```diff
@@ -184,4 +184,23 @@ void main() {
+    peer = new json_rpc.Peer.withoutJson(
+        new StreamChannel(incomingController.stream, outgoingController),
+        onUnhandledError: (error, stack) {
+      completer.complete(error);
```
Nit: need a little more indentation (`completer.complete` should be inside and more indented than `onUnhandledError: `).

dnfield replied:
That's how dartfmt did it

kevmoo replied:
Add a comma between `}` and `);` below and reformat...

dnfield replied:
Done

liyuqian replied:
Ah, this is magic. Is the old indentation a feature or an unexpected behavior of dartfmt?

kevmoo replied:
this is an...uh...obscure feature of dartfmt 😄

dnfield commented:
@nex3 @kevmoo is this good to land?  I don't have commit access here and I'm not sure if @liyuqian's LGTM is enough to commit here either

nex3 commented:
As I said, I no longer use or maintain this package, so I'm not the right person to ask :woman_shrugging:.

kevmoo commented:
@jacob314 – can you run w/ this?

liyuqian commented:
Any update? This would be very helpful for Flutter to debug its flaky tests. Currently we probably spent 100 hours of machine time and 10 hours of engineering time on those flaky tests every day.

kevmoo commented:
@vsmenon @dgrove @mit-mit – thoughts here?

kevmoo commented:
@natebosch ?

@liyuqian – please rebase against latest.

vsmenon commented:
This looks fine to me.

natebosch commented on CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 2.0.10
+
+* `Server` and related classes can now take an `onUnhandledError` callback to
```
This is a new feature for clients to use. The version should be bumped to `2.1.0`

dnfield replied:
Done

natebosch commented on lib/src/peer.dart:
```diff
@@ -39,13 +39,17 @@ class Peer implements Client, Server {
   /// Note that the peer won't begin listening to [channel] until [Peer.listen]
   /// is called.
-  Peer(StreamChannel<String> channel)
+  Peer(StreamChannel<String> channel, {ErrorCallback onUnhandledError})
```
This argument should be called out in the doc comment.

dnfield replied:
Done

natebosch commented on lib/src/server.dart:
```diff
@@ -15,6 +15,9 @@ import 'exception.dart';
 import 'utils.dart';
 
+/// A callback for unhandled exceptions.
+typedef ErrorCallback = void Function(dynamic error, dynamic stackTrace);
```
I'm not sure it's worth the `typedef` here. `void Function(dynamic, dynamic)` isn't that long.

dnfield replied:
I prefer keeping the `typedef` since it allows for documentation and naming of the parameters.  This callback is also repeated across several functions in peer.dart and server.dart.

If that really doesn't fit with the style here, I can certainly take it out - let me know.

kevmoo commented on CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 2.0.10
```
Since this adds a feature, it should be versioned as `2.1.0`

kevmoo replied:
...update the pubspec accordingly!

dnfield replied:
Done

liyuqian commented:
Thanks for the review! Unfortunately, @dnfield is OOO this week (sorry that I didn't check the calendar before pinging this). We'll probably wait until next week to have an updated patch for this.

natebosch commented:
Merged and published.
